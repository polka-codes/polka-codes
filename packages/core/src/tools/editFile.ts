// Tool for flexible file editing with before/after text anchors
// generated by polka.codes

import { type FullToolInfo, PermissionLevel, type ToolHandler, type ToolInfo, ToolResponseType } from '../tool'
import type { FilesystemProvider } from './provider'
import { getArray, getString } from './utils'
import { END_OF_FILE, type EditOperation, START_OF_FILE, editFile } from './utils/editFile'

export const toolInfo = {
  name: 'edit_file',
  description:
    'Request to edit file contents using before/after text anchors with flexible operations. Supports multiple edit operations in a single call.',
  parameters: [
    {
      name: 'path',
      description: 'The path of the file to edit',
      required: true,
      usageValue: 'File path here',
    },
    {
      name: 'operations',
      description: 'Edit operation with start_anchor, end_anchor, new_text, and optional line range hints',
      required: true,
      allowMultiple: true,
      children: [
        {
          name: 'start_anchor',
          description: `Text to find as the start anchor (use ${START_OF_FILE} for file start)`,
          required: false,
          usageValue: 'Text before the edit location',
        },
        {
          name: 'end_anchor',
          description: `Text to find as the end anchor (use ${END_OF_FILE} for file end)`,
          required: false,
          usageValue: 'Text after the edit location',
        },
        {
          name: 'new_text',
          description: 'Text to replace the content between start_anchor and end_anchor',
          required: true,
          usageValue: 'New text content',
        },
        {
          name: 'start_anchor_line_start',
          description: 'Optional line number hint for start_anchor location (1-based)',
          required: false,
          usageValue: '10',
        },
        {
          name: 'end_anchor_line_start',
          description: 'Optional line number hint for end_anchor location (1-based)',
          required: false,
          usageValue: '20',
        },
      ],
      usageValue: 'operations here',
    },
  ],
  examples: [
    {
      description: 'Replace content between two text anchors',
      parameters: [
        {
          name: 'path',
          value: 'src/main.ts',
        },
        {
          name: 'operations',
          value: {
            start_anchor: 'function oldFunction() {',
            end_anchor: '}',
            new_text: '\n  return "new implementation";\n',
          },
        },
      ],
    },
    {
      description: 'Insert at start of file',
      parameters: [
        {
          name: 'path',
          value: 'src/header.ts',
        },
        {
          name: 'operations',
          value: {
            start_anchor: START_OF_FILE,
            end_anchor: 'export',
            new_text: '// File header comment\n',
          },
        },
      ],
    },
    {
      description: 'Multiple operations in one call',
      parameters: [
        {
          name: 'path',
          value: 'src/utils.ts',
        },
        {
          name: 'operations',
          value: [
            {
              start_anchor: 'import React',
              end_anchor: 'from "react"',
              new_text: ', { useState }',
            },
            {
              start_anchor: 'function Component() {',
              end_anchor: 'return (',
              new_text: '\n  const [state, setState] = useState(false);\n  ',
            },
          ],
        },
      ],
    },
  ],
  permissionLevel: PermissionLevel.Write,
} as const satisfies ToolInfo

export const handler: ToolHandler<typeof toolInfo, FilesystemProvider> = async (provider, args) => {
  if (!provider.readFile || !provider.writeFile) {
    return {
      type: ToolResponseType.Error,
      message: 'Not possible to edit file. Abort.',
    }
  }

  const path = getString(args, 'path')
  const operations = getArray(args, 'operations') as unknown as EditOperation[]

  if (!operations || operations.length === 0) {
    return {
      type: ToolResponseType.Error,
      message: `<error><edit_file_path>${path}</edit_file_path><error_message>At least one edit operation is required</error_message></error>`,
    }
  }

  const fileContent = await provider.readFile(path)

  if (fileContent == null) {
    return {
      type: ToolResponseType.Error,
      message: `<error><edit_file_path>${path}</edit_file_path><error_message>File not found</error_message></error>`,
    }
  }

  try {
    const result = await editFile(fileContent, operations)
    await provider.writeFile(path, result)

    return {
      type: ToolResponseType.Reply,
      message: `<edit_file_path>${path}</edit_file_path>`,
    }
  } catch (error) {
    return {
      type: ToolResponseType.Error,
      message: `<error><edit_file_path>${path}</edit_file_path><error_message>${error instanceof Error ? error.message : String(error)}</error_message></error>`,
    }
  }
}

export const isAvailable = (provider: FilesystemProvider): boolean => {
  return !!provider.readFile && !!provider.writeFile
}

export default {
  ...toolInfo,
  handler,
  isAvailable,
} satisfies FullToolInfo
