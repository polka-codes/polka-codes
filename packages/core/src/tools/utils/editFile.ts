// Core logic for flexible file editing with before/after text anchors
// generated by polka.codes

/**
 * NOTE: Both `start_anchor` and `end_anchor` are **exclusive** markers.
 * They remain in the file; `new_text` is inserted *between* them.
 */
export interface EditOperation {
  start_anchor?: string
  end_anchor?: string
  new_text: string
  start_anchor_line_start?: number
  end_anchor_line_start?: number
}

export const START_OF_FILE = '<<<START_OF_FILE>>>'
export const END_OF_FILE = '<<<END_OF_FILE>>>'

/**
 * Apply multiple edit operations to file content using start/end text anchors (anchors are **exclusive** â€“ they stay in the file).
 * @param fileContent - The original file content
 * @param operations - Array of edit operations to apply
 * @returns The modified file content
 */
export const editFile = async (fileContent: string, operations: EditOperation[]): Promise<string> => {
  if (!operations || operations.length === 0) {
    throw new Error('At least one edit operation is required')
  }

  const originalLines = fileContent.split('\n')
  let updatedContent = fileContent

  // Process operations in order
  for (const operation of operations) {
    updatedContent = await applyEditOperation(updatedContent, operation, originalLines)
  }

  return updatedContent
}

async function applyEditOperation(fileContent: string, operation: EditOperation, originalLines: string[]): Promise<string> {
  const { start_anchor, end_anchor, new_text, start_anchor_line_start, end_anchor_line_start } = operation

  // Handle special cases for start/end of file
  if (start_anchor === START_OF_FILE && end_anchor === END_OF_FILE) {
    // Replace entire file
    return new_text
  }

  if (start_anchor === START_OF_FILE) {
    // Insert at start or replace from start to end_anchor
    if (!end_anchor) {
      return new_text + fileContent
    }
    const afterIndex = findTextWithHint(fileContent, end_anchor, end_anchor_line_start, originalLines)
    return new_text + fileContent.slice(afterIndex)
  }

  if (end_anchor === END_OF_FILE) {
    // Insert at end or replace from start_anchor to end
    if (!start_anchor) {
      return fileContent + new_text
    }
    const beforeIndex = findTextWithHint(fileContent, start_anchor, start_anchor_line_start, originalLines)
    const beforeEndIndex = beforeIndex + start_anchor.length
    return fileContent.slice(0, beforeEndIndex) + new_text
  }

  // Handle normal case with both start_anchor and end_anchor
  if (start_anchor && end_anchor) {
    const beforeIndex = findTextWithHint(fileContent, start_anchor, start_anchor_line_start, originalLines)
    const beforeEndIndex = beforeIndex + start_anchor.length
    const afterIndex = findTextWithHint(fileContent, end_anchor, end_anchor_line_start, originalLines, beforeEndIndex)

    return fileContent.slice(0, beforeEndIndex) + new_text + fileContent.slice(afterIndex)
  }

  // Handle case with only start_anchor (insert after)
  if (start_anchor) {
    const beforeIndex = findTextWithHint(fileContent, start_anchor, start_anchor_line_start, originalLines)
    const beforeEndIndex = beforeIndex + start_anchor.length
    return fileContent.slice(0, beforeEndIndex) + new_text + fileContent.slice(beforeEndIndex)
  }

  // Handle case with only end_anchor (insert before)
  if (end_anchor) {
    const afterIndex = findTextWithHint(fileContent, end_anchor, end_anchor_line_start, originalLines)
    return fileContent.slice(0, afterIndex) + new_text + fileContent.slice(afterIndex)
  }

  throw new Error('Either start_anchor or end_anchor must be specified')
}

function findTextWithHint(
  content: string,
  searchText: string,
  lineHint: number | undefined,
  originalLines: string[],
  startIndex = 0,
): number {
  // If we have a line hint, try searching around that area first
  if (lineHint && lineHint > 0 && lineHint <= originalLines.length) {
    const hintIndex = getLineStartIndex(originalLines, lineHint - 1) // Convert to 0-based
    const searchRadius = 5 // Search within 5 lines of the hint

    // Search in a window around the hint
    const windowStart = Math.max(0, hintIndex - searchRadius * 50) // Approximate chars per line
    const windowEnd = Math.min(content.length, hintIndex + searchRadius * 50)
    const windowContent = content.slice(windowStart, windowEnd)

    const relativeIndex = windowContent.indexOf(searchText)
    if (relativeIndex !== -1) {
      const absoluteIndex = windowStart + relativeIndex
      if (absoluteIndex >= startIndex) {
        return absoluteIndex
      }
    }
  }

  // Fallback to regular search from startIndex
  const index = content.indexOf(searchText, startIndex)
  if (index === -1) {
    throw new Error(`Could not find text: ${searchText}`)
  }
  return index
}

function getLineStartIndex(lines: string[], lineIndex: number): number {
  let index = 0
  for (let i = 0; i < lineIndex && i < lines.length; i++) {
    index += lines[i].length + 1 // +1 for newline character
  }
  return index
}
