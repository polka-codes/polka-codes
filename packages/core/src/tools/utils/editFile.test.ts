// Tests for editFile utility function
// generated by polka.codes

import { describe, expect, it } from 'bun:test'

import { END_OF_FILE, type EditOperation, START_OF_FILE, editFile } from './editFile'

describe('editFile', () => {
  const sampleContent = `line1
line2
line3
line4
line5`

  describe('basic operations', () => {
    it('should insert text after start_anchor', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line2',
          new_text: '\ninserted line',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should insert text before end_anchor', async () => {
      const operations: EditOperation[] = [
        {
          end_anchor: 'line3',
          new_text: 'inserted line\n',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should replace text between start_anchor and end_anchor', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line2',
          end_anchor: 'line4',
          new_text: '\nreplaced content\n',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should handle multiple operations in sequence', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line1',
          new_text: '\nafter line1',
        },
        {
          end_anchor: 'line5',
          new_text: 'before line5\n',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })
  })

  describe('special markers', () => {
    it('should replace entire file with START_OF_FILE and END_OF_FILE', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: START_OF_FILE,
          end_anchor: END_OF_FILE,
          new_text: 'completely new content',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toBe('completely new content')
    })

    it('should insert at start of file with START_OF_FILE', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: START_OF_FILE,
          new_text: 'new start\n',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should insert at end of file with END_OF_FILE', async () => {
      const operations: EditOperation[] = [
        {
          end_anchor: END_OF_FILE,
          new_text: '\nnew end',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should replace from START_OF_FILE to specific text', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: START_OF_FILE,
          end_anchor: 'line3',
          new_text: 'new beginning\n',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should replace from specific text to END_OF_FILE', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line3',
          end_anchor: END_OF_FILE,
          new_text: '\nnew ending',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })
  })

  describe('line hints', () => {
    const contentWithDuplicates = `line1
duplicate
line3
duplicate
line5`

    it('should use line hint to find correct occurrence', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'duplicate',
          end_anchor_line_start: 4, // Second occurrence
          end_anchor: 'line5',
          new_text: '\nfound second duplicate\n',
        },
      ]

      const result = await editFile(contentWithDuplicates, operations)
      expect(result).toMatchSnapshot()
    })

    it('should fallback to regular search when hint fails', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'duplicate',
          start_anchor_line_start: 999, // Invalid hint
          new_text: '\nfound duplicate',
        },
      ]

      const result = await editFile(contentWithDuplicates, operations)
      expect(result).toMatchSnapshot()
    })

    it('should use start_anchor_line_start hint', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'duplicate',
          start_anchor_line_start: 4, // Second occurrence
          new_text: '\nafter second duplicate',
        },
      ]

      const result = await editFile(contentWithDuplicates, operations)
      expect(result).toMatchSnapshot()
    })
  })

  describe('edge cases', () => {
    it('should handle empty file', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: START_OF_FILE,
          new_text: 'new content',
        },
      ]

      const result = await editFile('', operations)
      expect(result).toBe('new content')
    })

    it('should handle empty new_text', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line2',
          end_anchor: 'line4',
          new_text: '',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should handle whitespace in search text', async () => {
      const contentWithSpaces = `  line1
    line2
  line3`

      const operations: EditOperation[] = [
        {
          start_anchor: '    line2',
          new_text: '\n    inserted',
        },
      ]

      const result = await editFile(contentWithSpaces, operations)
      expect(result).toMatchSnapshot()
    })

    it('should handle newlines in search text', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line2\nline3',
          new_text: '\nreplaced multiline',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should handle single line file', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'single',
          new_text: ' modified',
        },
      ]

      const result = await editFile('single line', operations)
      expect(result).toBe('single modified line')
    })
  })

  describe('error cases', () => {
    it('should throw error when no operations provided', async () => {
      expect(editFile(sampleContent, [])).rejects.toThrow('At least one edit operation is required')
    })

    it('should throw error when operations is null', async () => {
      expect(editFile(sampleContent, null as any)).rejects.toThrow('At least one edit operation is required')
    })

    it('should throw error when start_anchor not found', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'nonexistent',
          new_text: 'new text',
        },
      ]

      expect(editFile(sampleContent, operations)).rejects.toThrow('Could not find text: nonexistent')
    })

    it('should throw error when end_anchor not found', async () => {
      const operations: EditOperation[] = [
        {
          end_anchor: 'nonexistent',
          new_text: 'new text',
        },
      ]

      expect(editFile(sampleContent, operations)).rejects.toThrow('Could not find text: nonexistent')
    })

    it('should throw error when neither start_anchor nor end_anchor specified', async () => {
      const operations: EditOperation[] = [
        {
          new_text: 'new text',
        },
      ]

      expect(editFile(sampleContent, operations)).rejects.toThrow('Either start_anchor or end_anchor must be specified')
    })

    it('should throw error when end_anchor appears before start_anchor', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line4',
          end_anchor: 'line2', // This appears before line4
          new_text: 'replacement',
        },
      ]

      expect(editFile(sampleContent, operations)).rejects.toThrow('Could not find text: line2')
    })
  })

  describe('complex scenarios', () => {
    it('should handle multiple operations with overlapping regions', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line1',
          new_text: '\nafter line1',
        },
        {
          start_anchor: 'line2',
          end_anchor: 'line3',
          new_text: '\nreplaced 2-3\n',
        },
        {
          end_anchor: 'line5',
          new_text: 'before line5\n',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should handle operations that create new searchable text', async () => {
      const operations: EditOperation[] = [
        {
          start_anchor: 'line2',
          new_text: '\nnew marker',
        },
        {
          start_anchor: 'new marker',
          new_text: '\nafter new marker',
        },
      ]

      const result = await editFile(sampleContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should handle code-like content with indentation', async () => {
      const codeContent = `function test() {
  const x = 1;
  const y = 2;
  return x + y;
}`

      const operations: EditOperation[] = [
        {
          start_anchor: 'const y = 2;',
          new_text: '\n  const z = 3;\n  const y = 2;',
        },
      ]

      const result = await editFile(codeContent, operations)
      expect(result).toMatchSnapshot()
    })

    it('should handle JSON-like content', async () => {
      const jsonContent = `{
  "name": "test",
  "version": "1.0.0",
  "dependencies": {}
}`

      const operations: EditOperation[] = [
        {
          start_anchor: '"dependencies": {}',
          new_text: '"scripts": {\n    "test": "bun test"\n  },\n  "dependencies": {}',
        },
      ]

      const result = await editFile(jsonContent, operations)
      expect(result).toMatchSnapshot()
    })
  })
})
