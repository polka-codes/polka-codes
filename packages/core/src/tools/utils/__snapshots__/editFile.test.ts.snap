// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`editFile basic operations should insert text after start_anchor 1`] = `
"line1
line2
inserted line
line3
line4
line5"
`;

exports[`editFile basic operations should insert text before end_anchor 1`] = `
"line1
line2
inserted line
line3
line4
line5"
`;

exports[`editFile basic operations should replace text between start_anchor and end_anchor 1`] = `
"line1
line2
replaced content
line4
line5"
`;

exports[`editFile basic operations should handle multiple operations in sequence 1`] = `
"line1
after line1
line2
line3
line4
before line5
line5"
`;

exports[`editFile special markers should insert at start of file with START_OF_FILE 1`] = `
"new start
line1
line2
line3
line4
line5"
`;

exports[`editFile special markers should insert at end of file with END_OF_FILE 1`] = `
"line1
line2
line3
line4
line5
new end"
`;

exports[`editFile special markers should replace from START_OF_FILE to specific text 1`] = `
"new beginning
line3
line4
line5"
`;

exports[`editFile special markers should replace from specific text to END_OF_FILE 1`] = `
"line1
line2
line3
new ending"
`;

exports[`editFile line hints should use line hint to find correct occurrence 1`] = `
"line1
duplicate
found second duplicate
line5"
`;

exports[`editFile line hints should fallback to regular search when hint fails 1`] = `
"line1
duplicate
found duplicate
line3
duplicate
line5"
`;

exports[`editFile line hints should use start_anchor_line_start hint 1`] = `
"line1
duplicate
after second duplicate
line3
duplicate
line5"
`;

exports[`editFile edge cases should handle empty new_text 1`] = `
"line1
line2line4
line5"
`;

exports[`editFile edge cases should handle whitespace in search text 1`] = `
"  line1
    line2
    inserted
  line3"
`;

exports[`editFile edge cases should handle newlines in search text 1`] = `
"line1
line2
line3
replaced multiline
line4
line5"
`;

exports[`editFile complex scenarios should handle multiple operations with overlapping regions 1`] = `
"line1
after line1
line2
replaced 2-3
line3
line4
before line5
line5"
`;

exports[`editFile complex scenarios should handle operations that create new searchable text 1`] = `
"line1
line2
new marker
after new marker
line3
line4
line5"
`;

exports[`editFile complex scenarios should handle code-like content with indentation 1`] = `
"function test() {
  const x = 1;
  const y = 2;
  const z = 3;
  const y = 2;
  return x + y;
}"
`;

exports[`editFile complex scenarios should handle JSON-like content 1`] = `
"{
  "name": "test",
  "version": "1.0.0",
  "dependencies": {}"scripts": {
    "test": "bun test"
  },
  "dependencies": {}
}"
`;
